### PR TITLE
Fullscan

### PR DESCRIPTION
**Pull request by bluefoxah**
add fullscan command for scan return with it's values.

### CODE REVIEW

File: server/merge.go [hunk 1]
```diff
@@ -30,22 +30,25 @@ func (s *Server) doMergeCommand(conn redcon.Conn, cmd redcon.Command) {
...
+		} else {
+			s.doMergeScan(conn, cmd)
+		}
 	}
 
 }
 
-func (s *Server) doMergeScan(conn redcon.Conn, cmd redcon.Command) {
+func (s *Server) doScanCommon(cmd redcon.Command) ([]interface{}, error) {
 	if scanJobCount >= s.maxScanJob {
```
Comment (absolute8511) [inline server/merge.go]: need atomic here

File: server/merge.go [hunk 2]
```diff
@@ -188,7 +295,7 @@ func (s *Server) decodeCursor(key []byte, nsBaseName string) (map[string]string,
 	nsMap := make(map[string]string)
 
 	if len(originCursor) == 2 {
-		table := string(originCursor[0])
+		table := originCursor[0]
```
Comment (absolute8511) [inline server/merge.go]: return early below if error to avoid too much indent

Comment (bluefoxah) [inline server/merge.go]: done

File: server/merge.go [hunk 3]
```diff
@@ -92,61 +91,169 @@ func (s *Server) doMergeScan(conn redcon.Conn, cmd redcon.Command) {
...
+
+func (s *Server) doMergeScan(conn redcon.Conn, cmd redcon.Command) {
+	results, err := s.doScanCommon(cmd)
+	if err != nil {
+		conn.WriteError(err.Error() + " : Err handle command " + string(cmd.Args[0]))
+		return
+	}
+
+	nextCursorBytes := []byte("")
+	result := make([]interface{}, 0)
```
Comment (bluefoxah) [inline server/merge.go]: done

File: rockredis/scan.go [hunk 4]
```diff
@@ -162,11 +188,14 @@ func (db *RockDB) scanGenericUseBuffer(storeDataType byte, key []byte, count int
 	}
 
 	for i := 0; it.Valid() && i < count; it.Next() {
-		if k, err := decodeScanKey(storeDataType, it.Key()); err != nil {
+		if t, k, err := decodeScanKey(storeDataType, it.Key()); err != nil {
 			continue
 		} else if r != nil && !r.Match(string(k)) {
 			continue
 		} else {
+			if !bytes.Equal(t, table) {
```
Comment (bluefoxah) [inline rockredis/scan.go]: done

File: rockredis/scan.go [hunk 5]
```diff
@@ -123,6 +143,7 @@ func decodeScanKey(storeDataType byte, ek []byte) (key []byte, err error) {
 	default:
 		err = errDataType
 	}
+	table, key, err = common.ExtraTable(key)
```
Comment (bluefoxah) [inline rockredis/scan.go]: done

File: server/merge.go (hunk 1 above)
Comment (bluefoxah) [inline server/merge.go]: done

File: rockredis/fullscan.go [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+			}
+		})
+}
+
+func (db *RockDB) fullScanCommon(tp byte, key []byte, count int, match string,
+	f itemFunc) *common.FullScanResult {
+	r, err := buildMatchRegexp(match)
+	if err != nil {
+		return &common.FullScanResult{
+			Results:    nil,
```
Comment (absolute8511) [inline rockredis/fullscan.go]: Maybe define a common FullScanResult error type to avoid duplicate code

File: rockredis/fullscan.go [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+			} else {
+				nextCursor, _ = encodeFullScanCursor(rk, container.cursor)
+			}
+		} else {
+			nextCursor, _ = encodeFullScanCursor(container.key, container.cursor)
+		}
+	}
+	return &common.FullScanResult{
+		Results:    result,
+		Type:       common.ZSET,
```
Comment (absolute8511) [inline rockredis/fullscan.go]: why type is ZSET here?

File: rockredis/fullscan.go [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+	} else {
+		if tp == KVType {
+			//cross table
+			//filter the cross table data
+			resLen := len(result)
+			if resLen > 0 {
+				item := result[resLen-1].([]interface{})
+				tab, _, err := common.ExtractTable(item[0].([]byte))
+				if err == nil && !bytes.Equal(tab, table) {
+					nextCursor, _ = encodeFullScanCursor([]byte(""), container.cursor)
```
Comment (absolute8511) [inline rockredis/fullscan.go]: I think the `nextCursor` here may be override by the below?

File: rockredis/fullscan.go [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+				return &common.FullScanResult{
+					Results:    nil,
+					Type:       common.ZSET,
+					NextCursor: nil,
+					PartionId:  "",
+					Error:      err,
+				}
+			}
+		}
+		tmp := tmpResult[string(container.key)]
```
Comment (absolute8511) [inline rockredis/fullscan.go]: optimize can be done since the key will be ordered

File: rockredis/fullscan.go (hunk 7 above)
Comment (bluefoxah) [inline rockredis/fullscan.go]: fix

File: rockredis/fullscan.go (hunk 8 above)
Comment (bluefoxah) [inline rockredis/fullscan.go]: done

File: rockredis/fullscan.go [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+		return db.setFullScan(key, count, match, inputBuffer)
+	case ZSetType:
+		return db.zsetFullScan(key, count, match, inputBuffer)
+	}
+	return &common.FullScanResult{
+		Results:    nil,
+		Type:       common.NONE,
+		NextCursor: nil,
+		PartionId:  "",
+		Error:      nil,
```
Comment (absolute8511) [inline rockredis/fullscan.go]: should return error with not supported scan type

File: rockredis/fullscan.go [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+			Type:       common.NONE,
+			NextCursor: nil,
+			PartionId:  "",
+			Error:      err,
+		}
+	}
+
+	table, rk, err := extractTableFromRedisKey(key)
+	if err != nil {
+		return &common.FullScanResult{
```
Comment (absolute8511) [inline rockredis/fullscan.go]: Maybe add  `NewFullScanResult(err *error)` to get a default FullScanResult and init with error(which can be nil)

File: rockredis/fullscan.go (hunk 6 above)
Comment (bluefoxah) [inline rockredis/fullscan.go]: done

File: rockredis/fullscan.go (hunk 10 above)
Comment (bluefoxah) [inline rockredis/fullscan.go]: done

File: rockredis/fullscan.go (hunk 11 above)
Comment (bluefoxah) [inline rockredis/fullscan.go]: done

File: rockredis/fullscan.go (hunk 9 above)
Comment (bluefoxah) [inline rockredis/fullscan.go]: done

File: rockredis/fullscan.go [hunk 12]
```diff
@@ -184,67 +187,56 @@ func (db *RockDB) fullScanCommon(tp byte, key []byte, count int, match string,
...
-			Type:       common.NONE,
-			NextCursor: nil,
-			PartionId:  "",
-			Error:      err,
-		}
+		return buildErrFullScanResult(err, common.NONE)
 	}
-	tmpResult := make(map[string][]interface{})
+	lock.Lock()
+	defer lock.Unlock()
```
Comment (absolute8511) [inline rockredis/fullscan.go]: What is this Lock for?

File: rockredis/fullscan.go [hunk 13]
```diff
@@ -184,67 +187,56 @@ func (db *RockDB) fullScanCommon(tp byte, key []byte, count int, match string,
...
-					Type:       common.NONE,
-					NextCursor: nil,
-					PartionId:  "",
-					Error:      err,
+				return buildErrFullScanResult(err, common.NONE)
+			}
+		}
+
+		if !bytes.Equal(prevKey, container.key) {
+			if !first {
```
Comment (absolute8511) [inline rockredis/fullscan.go]: I think the first is not needed here. 

```
if len(item) > 0 {
 				result = append(result, item)
 }
item = item[:0]
item = append(item, container.key)
```

**Pull request merged**